Add Series.dot method to dataframe module

Madhu94 opened the pull request:
Ref: #1259

- [ ] Closes #1259
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

I had a few questions about how to attempt this so I am marking this `WIP`. 

1. Should I handle the case for dot product of a dask series and a dask array? I didn't do so because a lot of the underlying infrastructure (like the `_extract_meta` utility) assumed the operands were all either dask series or dask dataframes.
2. Should I implement this for dask dataframe as part of this PR too or can that be done separately?
3. Should the method accept a `meta` kwarg too, for the case when the other operand could be a dataframe?

Any other feedback on the code would really help, thanks in advance!

jsignell commented:
Thanks for opening this! I'll give you my opinionated answers :)

> 1. Should I handle the case for dot product of a dask series and a dask array? I didn't do so because a lot of the underlying infrastructure (like the _extract_meta utility) assumed the operands were all either dask series or dask dataframes.

I think for this first PR it is ok to not accepst dask.arrays. It is a good idea though, to try to alert the user in a friendly error message if they do pass an array.

> 2. Should I implement this for dask dataframe as part of this PR too or can that be done separately?

It can be done separately. It'll make this easier to review if it just does series. And then the dataframe PR will be straightforward.

> 3. Should the method accept a meta kwarg too, for the case when the other operand could be a dataframe?

I think all methods should accept a meta kwarg.

Madhu94 commented:
I think I've addressed the above issues, thank you!  I had a query about the `enforce_metadata` kwarg accepted by `map_partitions` - it doesn't seem to enforce `dtype`, only the structure (i.e. `name` attribute in case of series or `columns` in case of dataframes). Is there a reason dtypes aren't enforced?

jsignell commented on dask/dataframe/tests/test_dataframe.py:
```diff
@@ -4502,3 +4502,46 @@ def test_assign_na_float_columns():
+    dask_df = dd.from_pandas(df, npartitions=1)
+    dask_s2 = dd.from_pandas(s2, npartitions=1)
+
+    assert_eq(s1.dot(s2), dask_s1.dot(dask_s2).compute())
+    assert_eq((s1.dot(df) == dask_s1.dot(dask_df).compute()).all(), True)
```
In all these tests that use `assert_eq` one argument should be the pandas object and the other should be the dask object:

```suggestion
    assert_eq(s1.dot(s2), dask_s1.dot(dask_s2))
    assert_eq(s1.dot(df) , dask_s1.dot(dask_df))
```

jsignell commented on dask/dataframe/core.py:
```diff
@@ -2573,6 +2573,27 @@ def append(self, other, interleave_partitions=False):
+    @derived_from(pd.Series)
+    def dot(self, other, meta=no_default):
+        if not isinstance(other, _Frame):
+            raise TypeError("The second operand must be dask array / dask dataframe")
```
```suggestion
            raise TypeError("The second argument must be a dask array or dask dataframe")
```

jsignell commented:
> I had a query about the `enforce_metadata` kwarg accepted by `map_partitions` - it doesn't seem to enforce `dtype`

It seems like `enforce_metadata` is intended to enforce dtype. Can you provide an example of that not working as expected?

Madhu94 commented:
> > I had a query about the `enforce_metadata` kwarg accepted by `map_partitions` - it doesn't seem to enforce `dtype`
> 
> It seems like `enforce_metadata` is intended to enforce dtype. Can you provide an example of that not working as expected?

```
import dask.dataframe as dd
import pandas as pd

s = pd.Series([1, 2, 3])
ds = dd.from_pandas(s, npartitions=1)
ds.map_partitions(lambda x: x.sum(), meta=float, enforce_metadata=True).compute()
```

I would expect this to coerce the result to float or raise an error. The reverse doesn't seem to work either -
```
s = pd.Series([1.1, 2.1, 3.1])
ds = dd.from_pandas(s, npartitions=1)
ds.map_partitions(lambda x: x.sum(), meta=int, enforce_metadata=True).compute()
```

Nothing in the code (`_apply_and_enforce` utility) seems to be checking `dtype`.  Let me know if I'm missing something?

Madhu94 commented:
The comment inside `_dot_series` is misleading actually, fixing that.

jsignell commented on dask/dataframe/tests/test_dataframe.py:
```diff
@@ -4502,3 +4502,46 @@ def test_assign_na_float_columns():
+    dask_df = dd.from_pandas(df, npartitions=1)
+    dask_s2 = dd.from_pandas(s2, npartitions=1)
+
+    assert_eq(s1.dot(s2), dask_s1.dot(dask_s2).compute())
+    assert_eq(s1.dot(df), dask_s1.dot(dask_df).compute())
```
It's better to not use `compute` in `assert_eq`. That way the `_meta` is also checked. 

```suggestion
    assert_eq(s1.dot(s2), dask_s1.dot(dask_s2))
    assert_eq(s1.dot(df), dask_s1.dot(dask_df))
```

Madhu94 replied:
I didn't know that, that is cool!

jsignell commented on dask/dataframe/tests/test_dataframe.py:
```diff
@@ -4502,3 +4502,46 @@ def test_assign_na_float_columns():
+    partitioned_df = dd.from_pandas(df, npartitions=2)
+    partitioned_s2 = dd.from_pandas(s2, npartitions=2)
+
+    assert_eq(s1.dot(s2), partitioned_s1.dot(partitioned_s2).compute())
+    assert_eq(s1.dot(df), partitioned_s1.dot(partitioned_df).compute())
```
```suggestion
    assert_eq(s1.dot(s2), partitioned_s1.dot(partitioned_s2))
    assert_eq(s1.dot(df), partitioned_s1.dot(partitioned_df))
```

jsignell commented on dask/dataframe/tests/test_dataframe.py:
```diff
@@ -4502,3 +4502,46 @@ def test_assign_na_float_columns():
+
+    # Test passing meta kwarg
+    res = dask_s1.dot(dask_df, meta=pd.Series([1], name="test_series")).compute()
+    assert_eq(res.name, "test_series")
```
Easier to use a regular assert here. 

```suggestion
    assert res.name ==  "test_series"
```

jsignell commented on dask/dataframe/tests/test_dataframe.py:
```diff
@@ -4502,3 +4502,46 @@ def test_assign_na_float_columns():
+        dask_s1.dot(da.array([1, 2, 3, 4]))
+
+
+def test_dot_nan():
```
It would be nice to check that these dask outputs match the equivalent in pandas.

jsignell replied:
sorry can you add some `assert_eq` to check the output against `s1.dot(s2)` for instance?

Madhu94 replied:
I went and manually checked the output and marked this as resolved - I don't know what I was thinking.

jsignell replied:
haha no worries :)

jsignell reviewed (commented):
Sorry I let this sit for a bit. I think it looks really nice now. I just have a few suggestions of how to improve the tests.

Madhu94 commented:
Checking CI

jsignell commented:
Thanks for sticking with this @Madhu94!

Madhu94 commented:
Thanks for your help @jsignell  :tada: :tada:

Madhu94 commented:
I believe  #1259 should not have been closed as a result of this since there were other methods to be implemented there. I'm sorry my first update wasn't clear. I'll be more careful going forward
